src/crypto/ecdh.rs

```diff
@@ -128,13 +182,72 @@ pub fn decrypt(priv_key: &ECDHSecretKey, mpis: &[Mpi], fingerprint: &[u8]) -> Re
 
     // Perform key derivation
     let z = kdf(
-        priv_key.hash,
+        priv_key.hash(),
         shared_secret.as_bytes(),
-        priv_key.alg_sym.key_size(),
-        &param,
+        priv_key.alg_sym().key_size(),
+        param,
     )?;
 
-    // Peform AES Key Unwrap
+    aes_key_unwrap(mpis, &z)
+}
+
+fn decrypt_p521(priv_key: &ECDHSecretKey, mpis: &[Mpi], param: &[u8]) -> Result<Vec<u8>> {
+    use cipher::Unsigned;
+    use p521::{
+        elliptic_curve::{ecdh, Curve, NonZeroScalar},
+        NistP521, PublicKey,
+    };
+
+    ensure_eq!(mpis.len(), 3);
+    // 66 = ((521 + 7) / 8)
+    ensure_eq!(
+        priv_key.secret().len(),
+        <NistP521 as Curve>::FieldBytesSize::USIZE,
+        "invalid secret point"
+    );
+
+    let Ok(their_public) = PublicKey::from_sec1_bytes(mpis[0].as_bytes()) else {
+        bail!("invalid public key");
+    };
+
+    let our_secret = GenericArray::from_slice(priv_key.secret()).to_owned();
+    let Some(private_scalar): Option<NonZeroScalar::<_>> = NonZeroScalar::<p521::NistP521>::from_repr(our_secret).into() else {
+        bail!("invalid private scalar");
+    };
+
+    let public_affine = their_public.as_affine();
+    let shared_secret = ecdh::diffie_hellman(private_scalar, public_affine);
+
+    // Perform key derivation
+    let z = kdf(
+        priv_key.hash(),
+        shared_secret.raw_secret_bytes().as_slice(),
+        priv_key.alg_sym().key_size(),
+        param,
+    )?;
+
+    aes_key_unwrap(mpis, &z)
+}
+
+/// Key Derivation Function for ECDH (as defined in RFC 6637).
+/// https://tools.ietf.org/html/rfc6637#section-7
```

Comment:
Invalid doc comment link

```suggestion
/// <https://tools.ietf.org/html/rfc6637#section-7>
```